Improvement of OpenCL device discovery and selection

voidxno opened the pull request:
PR for fix of OpenCL device problems described in **[issue #268](https://github.com/madMAx43v3r/mmx-node/issues/268)**

_NOTE: **Only review/feedback at this stage** (complete try that is working though)_

**Goal:** Solve problem of identical platform, or device names
**Goal:** Legacy/migrate/support existing config value logic
**Goal:** Single WebGUI selection list (all devices, all platforms)
**Goal:** Possible to find/configure OpenCL device without WebGUI
**Goal:** Keep logic simple

Logic (new):
- Enum all OpenCL devices, on all platforms, in a list, log(INFO)
- If `opencl_device` is `-1`, overrides everything, 'None' (CPU)
- If no OpenCL devices, fallback to 'None' (CPU)
- If `opencl_device_name` is defined (new logic)
  - Search list `opencl_device_name` hits, independent of platform, match if counter equals `opencl_device`
  - If not found, log(WARN), fallback to 'None' (CPU)
- If `opencl_device_name` empty (hybrid/old logic)
  - If `platform` (name) empty, define it temporarily to [0] platform
  - Search list for devices with `platform` name, match if counter equals `opencl_device`
  - If not found, log(WARN), fallback to 'None' (CPU)

Comments:
- All comments in code will be removed before any final merge/PR
- Existing .json config never changed, until selection change in WebGUI
- 'First device found' eliminated visually, still works behind the scenes
- log(INFO) of discovered OpenCL devices, makes manual .json config possible

<ins>**Questions**:</ins>
- Anything you observe/want adjusted, valid, just give feedback
- ~~I have a problem with keep-it-simple and `Node.opencl_device_selidx` value (written to .json):~~
  - ~~Want a persistent/temporary value in WebGUI, on browser reload, without permanent .json write~~
  - ~~If not, will have to duplicate 'device selection' logic from Node.cpp into settings.js (don't want to)~~
- Added a new 'Reward' separator split/card/header in WebGUI (just a suggestion)

Log (sample):
```
[Node] INFO: Found OpenCL GPU device 'NVIDIA GeForce RTX 3050' [0] (NVIDIA CUDA)
[Node] INFO: Found OpenCL GPU device 'NVIDIA GeForce GT 1030' [0] (NVIDIA CUDA)
[Node] INFO: Found OpenCL GPU device 'Intel(R) Graphics [0x7d67]' [0] (Intel(R) OpenCL Graphics)
[Node] INFO: Using OpenCL GPU device 'NVIDIA GeForce RTX 3050' [0] (NVIDIA CUDA)
```

WebGUI (sample):
```
OpenCL Device (selection list):
- None
- NVIDIA GeForce RTX 3050
- NVIDIA GeForce GT 1030
- Intel(R) Graphics [0x7d67]
```

voidxno commented:
Ref. my `Node.opencl_device_selidx` issue above. Looked at:
- `this.set_config()` call chain (`settings.js`)

One possibility:
- Add an extra boolean param to `this.set_config()`
- Indicates if write to .json/file should be performed
- `"/config/set"` (`WebAPI.cpp`)
  - `false`: Only perform `vnx::set_config`
  - `true`: Perform both `vnx::set_config` and `vnx::write_config_file` (todays default)

Not sure if right/best way. Or correct usage of vnx framework, as implemented now.

voidxno commented:
`Rev02` update:
- Implementation of boolean param to `this.set_config()` (true/false write to .json)
- Eliminated `Node.opencl_device_selidx`, by using existing `Node.opencl_device`

madMAx43v3r commented:
need to resolve some conflicts now

voidxno commented:
`Rev03` update:
- Resolved conflicts (i.e., redone with current mainnet-rc `master` code)

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+			log(INFO) << "No OpenCL device used (disabled)";
+		}
+
+		vnx::write_config("Node.opencl_device", listsel); // NOTE: 'Hijack' Node.opencl_device temporary to communicate 'OpenCL Device' selection WebGUI
```
What you mean by temporary?

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+					if( automy::basic_opencl::get_device_name(idd) == list[i].device_name ) device_index++;
+				}
+				list.push_back({automy::basic_opencl::get_device_name(idd), device_index, automy::basic_opencl::get_platform_name(idp), idp, idd});
+				log(INFO) << "Found OpenCL GPU device '" << list[list.size() - 1].device_name << "' [" << list[list.size() - 1].device_index << "] (" << list[list.size() - 1].platform_name << ")";
```
`list[list.size() - 1]` is the same as `list.back()`

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+			const auto devices = automy::basic_opencl::get_devices(idp, CL_DEVICE_TYPE_GPU);
+			for(const auto idd : devices) {
+				int device_index = 0;
+				for(size_t i = 0; i < list.size(); ++i) {
```
for loops like this it's better to use C++11 syntax:
```
const auto name = automy::basic_opencl::get_device_name(idd);
for(const auto& device : list) {
  if(device.device_name == name) {
    device_index++;
  }
}
```

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
-			for(const auto id : platforms) {
-				list.push_back(automy::basic_opencl::get_platform_name(id));
+
+		struct device { // NOTE: Element for each OpenCL device found
```
to keep it less verbose, since the struct is called `device`:
```
struct device_t {
  int index;
  std::string name;
  std::string platform;     // name
  ...
}
```

madMAx43v3r replied:
oh and always initialize primitives to some default value. `int index = 0;` in this case.

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("opencl.platform", "", true, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_index[value] : -1, true, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
+				this.set_config("Node.opencl_device", value, false, true); // NOTE: 'Hijack' value temporary for device selection WebGUI (without config file write), browser reload
```
I get it now, you can just invent a new config variable to store something in RAM.

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+			device_list_name.push_back(list[i].device_name);
+			device_list_index.push_back(list[i].device_index);
 		}
+		vnx::write_config("Node.device_list_name", device_list_name);
```
I think it's better to use a single `std::vector<std::pair<std::string, int>>` for this. And a single config as well.

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
 		opencl_device(value, prev) {
 			if(prev != null) {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("opencl.platform", "", true, true); // NOTE: Config change made, force any existing platform name config to ""
```
let's set it to `null`, will yield and empty string as well

voidxno replied:
Thx, done in `Rev04` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
+				this.set_config("opencl.platform", "", true, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_index[value] : -1, true, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
+				this.set_config("Node.opencl_device", value, false, true); // NOTE: 'Hijack' value temporary for device selection WebGUI (without config file write), browser reload
+				this.set_config("Node.opencl_device_name", (value >= 0) ? this.opencl_device_list_name[value] : "", true, true); // NOTE: Device name last, visually shown 'restart needed to apply' (WebGUI), usually (threading)
```
same here, set to `null` in fallback

voidxno replied:
Thx, done in `Rev04` update

voidxno commented:
`Rev04` update:
- Resolved comments from `Rev03`
  - `list.back()` instead of `list[list.size() - 1]`
  - foreach C++11 loops where possible
  - renamed struct to `device_t`, less verbose, init primitive `0`
  - new RAM config variable `Node.opencl_device_select`
  - single `std::vector<std::pair<std::string, int>>` usage
  - .json `""` config values changed to `null`

NOTE: If/when get to a result that looks workable:
- Will remove all comments in code
- Do final test of all scenarios on my test machines

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+						device_index++;
+					}
+				}
+				list.push_back({device_index, automy::basic_opencl::get_device_name(idd), automy::basic_opencl::get_platform_name(idp), idd, idp});
```
can reuse `name` here, and I would store platform name outside the loop, so you don't call `automy::basic_opencl::get_platform_name` as often.

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+					}
+				}
+				list.push_back({device_index, automy::basic_opencl::get_device_name(idd), automy::basic_opencl::get_platform_name(idp), idd, idp});
+				log(INFO) << "Found OpenCL GPU device '" << list.back().name << "' [" << list.back().index << "] (" << list.back().platform << ")";
```
you can allocate the device on stack first, then use the stack variable here and push it. this way the code is cleaner as you dont query the data from the vector.

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
-			platform = automy::basic_opencl::find_platform_by_name(platform_name);
+		std::vector<std::pair<std::string, int>> device_list;
+		for(const auto& device : list) {
+			device_list.push_back(std::make_pair(device.name, device.index));
```
another C++11 trick here: `device_list.emplace_back(device.name, device.index)`

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on src/Node.cpp:
```diff
@@ -52,55 +52,83 @@ void Node::main()
+					}
+				}
+				if(listsel >= 0) { // NOTE: Found OpenCL device in list to use
+					opencl_context = automy::basic_opencl::create_context(list[listsel].platform_id, {list[listsel].device_id});
```
since you use `list[listsel]` many times, it's cleaner to setup an alias for it: `const auto& device = list[listsel];`
the only catch is you cannot modify `list` while using such a reference, but you're not doing it here so it's fine.

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on src/WebAPI.cpp:
```diff
@@ -1138,37 +1138,41 @@ void WebAPI::http_request_async(std::shared_ptr<const vnx::addons::HttpRequest>
 		const auto iter_key = args.field.find("key");
 		const auto iter_value = args.field.find("value");
-		if(iter_key != args.field.end() && iter_value != args.field.end())
+		const auto iter_file = args.field.find("file"); // NOTE: Boolean option to write or not to config file (.json)
```
this param should be optional and default to `true`.

madMAx43v3r replied:
the name `file` is a bit misleading too, it sounds like a file name. but since you want a special case, the param should be inverted, so it's true when you want the special case. it could then be named `tmp_only` for example.

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
-					this.opencl_platform = data["opencl.platform"] != null ? data["opencl.platform"] : "";
-					this.opencl_device = data["Node.opencl_device"] != null ? data["Node.opencl_device"] : 0;
-					this.opencl_device_list = [{name: "None", value: -1}];
+					this.opencl_device = data["Node.opencl_device_select"] != null ? data["Node.opencl_device_select"] : -1; // NOTE: Current List selection (not relative, value index in WebGUI list), from temporary config (RAM only) value, not value written to config .json (to support refresh of WebGUI browser)
```
some JS magic to make it cleaner:
`this.opencl_device = data["Node.opencl_device_select"] ?? -1;`

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
 				});
 		},
-		set_config(key, value, restart) {
+		set_config(key, value, file, restart) { // NOTE: Added boolean option to write or not to config file (.json)
```
and put the new param at the end, with a default value, so you don't have to change every call

voidxno replied:
Thx, done in `Rev05` update

voidxno commented:
`Rev05` update:
- Resolved comments from `Rev04`
  - reused `name`, less calls to `automy::basic_opencl::get_platform_name`
  - didn't allocate device on stack, but used already established variables (looked clean/efficient)
  - used `.emplace_back` for `device_list` (nice trick)
  - did the `const auto& device = list[listsel];` (cleaner look)
  - reworked `file` to an optional `tmp_only` param (default `false`, inverted logic from before)
  - used `this.opencl_device = data["Node.opencl_device_select"] ?? -1;` (neat trick)
  - reworked `set_config()` with optional `tmp_only` param at end (default `false`, inverted logic from before)

**Please check:**
- The optional default `tmp_only` logic in `settings.js` (`set_config()`)
- The optional default `tmp_only` logic in `WebAPI.cpp` (`/config/set`)
- Basically:
  - Made both areas handle `tmp_only` optionally (independent of each other)
  - Not 100% sure/happy with `const bool tmp_only = ...` in `WebAPI.cpp`
- It all works, but it is 'save config' logic, do not want a bug there

_BTW: Thanks for patience/corrections. It's a good education in C++ :-) Starting to see elements of how you manage such an efficient (and large) codebase._

madMAx43v3r commented on src/WebAPI.cpp:
```diff
@@ -1138,37 +1138,41 @@ void WebAPI::http_request_async(std::shared_ptr<const vnx::addons::HttpRequest>
-				const auto pos = key.find('.');
-				if(pos == std::string::npos) {
-					file = key;
+			const bool tmp_only = iter_tmp_only != args.field.end() ? (iter_tmp_only->second.is_bool() ? iter_tmp_only->second : false) : false;
```
Since `args` is a `vnx::Object` and we don't require the param to be set, this can be simplified to:
```
const bool tmp_only = args["tmp_only"];
```

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
 				.then(response => {
 					if(response.ok) {
-						this.result = {key: req.key, value: req.value, restart};
+						this.result = {key: req.key, value: req.value, restart, tmp_only: req.tmp_only};
```
not really needed

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
 				});
 		},
-		set_config(key, value, restart) {
+		set_config(key, value, restart, tmp_only) { // NOTE: Added optional boolean option to write or not to config file (.json), default 'false'
```
`set_config(key, value, restart, tmp_only = false)`

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
-							for(const name of list) {
-								this.opencl_device_list.push({name: name, value: i++});
+							for(const device of list) {
+								this.opencl_device_list_relidx.push({name: device[0], index: device[1]});
```
Looks to me like you only need `index` from this array? no need for the object

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -157,14 +145,12 @@ Vue.component('node-settings', {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("Node.opencl_device_select", value, true, true); // NOTE: Temporary current device selection WebGUI (no config file write), survive browser reload
+				this.set_config("opencl.platform", null, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_relidx[value]["index"] : -1, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
```
```
const device = this.opencl_device_list_relidx[value];
```
And then get the fields via `device.index` and `device.name`.

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -157,14 +145,12 @@ Vue.component('node-settings', {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("Node.opencl_device_select", value, true, true); // NOTE: Temporary current device selection WebGUI (no config file write), survive browser reload
+				this.set_config("opencl.platform", null, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_relidx[value]["index"] : -1, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
```
`this.opencl_device_list_relidx[value].index` and `this.opencl_device_list_relidx[value].name`

voidxno replied:
Thx, done in `Rev05` update

madMAx43v3r commented on www/web-gui/public/settings.js:
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
 							let i = 0;
-							for(const name of list) {
-								this.opencl_device_list.push({name: name, value: i++});
+							for(const device of list) {
```
`for(const [i, device] of list.entries())`

voidxno replied:
Don't see a way to make work. Loop **inits 2x lists**:
`this.opencl_device_list_relidx.push({name: device[0], index: device[1]});`
`this.opencl_device_list.push({name: device[0], value: i++});`
**First one:** Absolute devices found, name and relative index (used to write to config).
**Second one:** Display list, name and selection index, including None/-1 (HTML drop-down).
**Second one:** `i` in this case just to init selection index value (`i++`).

voidxno replied:
**Maybe alternative** (cleaner/more understandable):

```
let selidx = 0;
for(const [name, relidx] of list.entries()) {
        this.opencl_device_list_relidx.push({name: name, index: relidx});
        this.opencl_device_list.push({name: name, value: selidx++});
}
```

madMAx43v3r replied:
oh no, I just wanted to replace the `i` with a builtin counter from the loop. `list.entries()` gives you both the value and the index in the array, so you don't need to count by yourself.

voidxno replied:
Now I get it, neat trick 👍 done in `Rev07` update

voidxno commented:
`Rev07` update:
- Resolved leftover **(except)** comment from `Rev05` 
  - neat trick with `for(const [i, device] of list.entries())` for `i` count

madMAx43v3r commented:
Looks good now, I tested it as well, works. But I only have one GPU to select.

madMAx43v3r commented:
The settings page needs to be re-worked with tabs, but in the meantime we can merge your change.

voidxno commented:
Suggestion:
- Let me make a last revision (remove all 'NOTE' comments), fix static 'Rewards' label/text
- Give me a day to retest different scenarious on my test machines (several GPU's, no GPU, so on)
- Then I'll reply when finished

voidxno commented:
Testing in progress (I'll reply when finished)

voidxno commented:
Testing done Linux (with latest `Rev08`):
- Machine with 2x Intel GPU / 2x NVIDIA GPU
- Machine with 1x AMD GPU
- Migrate of existing config value logic
- Identical platform, or device names
- Quick test with Windows build

Merge ok from my side.
